validator/client/propose.go

```diff
@@ -151,6 +161,122 @@ func (v *validator) ProposeBlock(ctx context.Context, slot types.Slot, pubKey [4
+	if err != nil {
+		log.WithError(err).Error("Failed to sign randao reveal")
+		if v.emitAccountMetrics {
+			ValidatorProposeFailVec.WithLabelValues(fmtKey).Inc()
+		}
+		return
+	}
+
+	g, err := v.getGraffiti(ctx, pubKey)
+	if err != nil {
+		// Graffiti is not a critical enough to fail block production and cause
+		// validator to miss block reward. When failed, validator should continue
+		// to produce the block.
+		log.WithError(err).Warn("Could not get graffiti")
+	}
+
+	// Request block from beacon node
+	b, err := v.validatorClient.GetBlockV2(ctx, &ethpb.BlockRequest{
+		Slot:         slot,
+		RandaoReveal: randaoReveal,
+		Graffiti:     g,
+	})
+	if err != nil {
+		log.WithField("blockSlot", slot).WithError(err).Error("Failed to request block from beacon node")
+		if v.emitAccountMetrics {
+			ValidatorProposeFailVec.WithLabelValues(fmtKey).Inc()
+		}
+		return
+	}
+
+	// Sign returned block from beacon node
+	sig, domain, err := v.signBlock(ctx, pubKey, epoch, interfaces.WrappedAltairBeaconBlock(b))
+	if err != nil {
+		log.WithError(err).Error("Failed to sign block")
+		if v.emitAccountMetrics {
+			ValidatorProposeFailVec.WithLabelValues(fmtKey).Inc()
+		}
+		return
+	}
+	blk := &ethpb.SignedBeaconBlockAltair{
+		Block:     b,
+		Signature: sig,
+	}
+
+	signingRoot, err := helpers.ComputeSigningRoot(b, domain.SignatureDomain)
+	if err != nil {
+		if v.emitAccountMetrics {
+			ValidatorProposeFailVec.WithLabelValues(fmtKey).Inc()
+		}
+		log.WithError(err).Error("Failed to compute signing root for block")
+		return
+	}
+
+	if err := v.preBlockSignValidations(ctx, pubKey, interfaces.WrappedAltairBeaconBlock(b), signingRoot); err != nil {
+		log.WithFields(
+			blockLogFields(pubKey, interfaces.WrappedAltairBeaconBlock(b), nil),
+		).WithError(err).Error("Failed block slashing protection check")
+		return
+	}
+
+	if err := v.postBlockSignUpdate(ctx, pubKey, interfaces.WrappedAltairSignedBeaconBlock(blk), signingRoot); err != nil {
+		log.WithFields(
+			blockLogFields(pubKey, interfaces.WrappedAltairBeaconBlock(b), sig),
+		).WithError(err).Error("Failed block slashing protection check")
+		return
+	}
+
+	// Propose and broadcast block via beacon node
+	blkResp, err := v.validatorClient.ProposeBlockV2(ctx, blk)
+	if err != nil {
+		log.WithError(err).Error("Failed to propose block")
+		if v.emitAccountMetrics {
+			ValidatorProposeFailVec.WithLabelValues(fmtKey).Inc()
+		}
+		return
+	}
+
+	span.AddAttributes(
+		trace.StringAttribute("blockRoot", fmt.Sprintf("%#x", blkResp.BlockRoot)),
+		trace.Int64Attribute("numDeposits", int64(len(b.Body.Deposits))),
+		trace.Int64Attribute("numAttestations", int64(len(b.Body.Attestations))),
+	)
+
+	blkRoot := fmt.Sprintf("%#x", bytesutil.Trunc(blkResp.BlockRoot))
+	log.WithFields(logrus.Fields{
+		"slot":            b.Slot,
+		"blockRoot":       blkRoot,
+		"numAttestations": len(b.Body.Attestations),
+		"numDeposits":     len(b.Body.Deposits),
+		"graffiti":        string(b.Body.Graffiti),
```

Comment:
Would it be useful to add a `version` field and log `altair`?

Reply:
Good idea, can add it in